pyomo/contrib/cp/repn/docplex_writer.py

```diff
@@ -0,0 +1,1094 @@
+
+    finalizeResult = None
+
+
+@WriterFactory.register(
+    'docplex_model', 'Generate the corresponding docplex model object')
+class DocplexWriter(object):
+    CONFIG = ConfigDict('docplex_model_writer')
+    CONFIG.declare('symbolic_solver_labels', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Write Pyomo Var and Constraint names to docplex model',
+    ))
+
+    def __init__(self):
+        self.config = self.CONFIG()
+
+    def write(self, model, **options):
+        config = options.pop('config', self.config)(options)
+
+        sorter = SortComponents.deterministic
+        component_map, unknown = categorize_valid_components(
+            model,
+            active=True,
+            sort=sorter,
+            valid={
+                Block, Objective, Constraint, Var, Param, BooleanVar,
+                LogicalConstraint, Suffix,
+                # FIXME: Non-active components should not report as Active
+                Set, RangeSet, Port,
+            },
+            targets={
+                Objective, Constraint, LogicalConstraint, IntervalVar
+            }
+        )
+        if unknown:
+            raise ValueError(
+                "The model ('%s') contains the following active components "
+                "that the docplex writer does not know how to process:\n\t%s" %
+                (model.name, "\n\t".join("%s:\n\t\t%s" % (
+                    k, "\n\t\t".join(map(attrgetter('name'), v)))
+                    for k, v in unknown.items())))
+
+        cpx_model = cp.CpoModel()
+        visitor = LogicalToDoCplex(
+            cpx_model,
+            symbolic_solver_labels=config.symbolic_solver_labels)
+
+        active_objs = []
+        for block in component_map[Objective]:
+            for obj in block.component_data_objects(Objective,
+                                                    sort=sorter,
+                                                    active=True,
+                                                    descend_into=False):
+                active_objs.append(obj)
+                # [ESJ 09/29/22]: TODO: I think that CP Optimizer can support
+                # multiple objectives. We should generalize this later, but for
+                # now I don't much care.
+                if len(active_objs) > 1:
+                    raise ValueError(
+                        "More than one active objective defined for "
+                        "input model '%s': Cannot write to docplex."
+                        % model.name)
+                obj_expr = visitor.walk_expression((obj.expr, obj, 0))
+                if obj.sense is minimize:
+                    cpx_model.add(cp.minimize(obj_expr[1]))
+                else:
+                    cpx_model.add(cp.maximize(obj_expr[1]))
+
+        # No objective is fine too, this is CP afterall...
+
+        # Write algebraic constraints
+        for block in component_map[Constraint]:
```

Comment:
I think we should move away from the `component_map` approach -- at least for Var/Constraint-like things: if we get all the constraints from a single `component_data_objects(Constraint)` call, then the function will remove duplicates (e.g., due to References).  If we have multiple calls to `component_data_objects`, then deduplication will only happen within each call and not between them.